hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenRejectHandler.java

```diff
@@ -0,0 +1,380 @@
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+/**
+ * This class contains all workflow-related functionality regarding {@link HederaFunctionality#TOKEN_REJECT}.
+ * This transaction type is used to reject tokens from an account and send them back to the treasury.
+ */
+@Singleton
+public class TokenRejectHandler extends BaseTokenHandler implements TransactionHandler {
+
+    @Inject
+    public TokenRejectHandler() {}
+
+    @Override
+    public void preHandle(@NonNull final PreHandleContext context) throws PreCheckException {
+        requireNonNull(context);
+        final var txn = context.body();
+        final var op = txn.tokenRejectOrThrow();
+
+        // If owner account is specified, verify that it's valid and require that it signs the transaction.
+        // If not specified, payer account is considered as the one rejecting the tokens.
+        if (op.hasOwner()) {
+            final var accountStore = context.createStore(ReadableAccountStore.class);
+            verifyOwnerAndRequireKey(op.ownerOrThrow(), context, accountStore);
+        }
+    }
+
+    /**
+     * Verifies that the owner account exists and ensures the account's key is valid and required for the transaction.
+     *
+     * @param ownerId The AccountID of the owner whose key needs to be validated.
+     * @param context The PreHandleContext providing transaction context.
+     * @param accountStore The store to access readable account information.
+     * @throws PreCheckException If the sender's account is immutable or the sender's account ID is invalid.
+     */
+    private void verifyOwnerAndRequireKey(
+            @NonNull final AccountID ownerId,
+            @NonNull final PreHandleContext context,
+            @NonNull final ReadableAccountStore accountStore)
+            throws PreCheckException {
+
+        final var ownerAccount = accountStore.getAliasedAccountById(ownerId);
+        validateTruePreCheck(ownerAccount != null, INVALID_OWNER_ID);
+
+        // If the sender account is immutable, then we throw an exception.
+        final var key = ownerAccount.key();
+        if (key == null || !isValid(key)) {
+            throw new PreCheckException(ACCOUNT_IS_IMMUTABLE);
+        }
+        context.requireKey(key);
+    }
+
+    @SuppressWarnings("java:S2259")
+    @Override
+    public void pureChecks(@NonNull final TransactionBody txn) throws PreCheckException {
+        requireNonNull(txn, "Transaction body cannot be null");
+        final var op = txn.tokenRejectOrThrow();
+
+        validateFalsePreCheck(op.rejections().isEmpty(), INVALID_TRANSACTION_BODY);
+        if (op.hasOwner()) {
+            validateAccountID(op.owner(), null);
+        }
+
+        final var uniqueTokenReferences = new HashSet<TokenReference>();
+        for (final var rejection : op.rejections()) {
```

Comment:
Should we also have a check we reject if each rejection has more than 2 account amount transfers ?

Reply:
Hey! 
Our only use case is:
Bob rejects token A he owns and the treasury account Carol receives the whole token A balance of Bob.

And since:
1 TokenTransfer = 1 TokenReference.
1 Token Transfer body contains List<AccountAmounts>

We always have exactly 2 AccountAmounts, because we have 2 different accounts that exchange tokens.

Reply:
@MrValioBg Since user can submit any invalid transactions, do we reject any transaction that is not like the valid use case ?

Reply:
Users can only specify the token they want to reject as a TokenReference. ( They don't specify treasury accounts, we retrieve that)
We have checks if they actually own it. If they don't have a balance of it, the TokenReject fails.